mcs/class/System/corefx/Interop.Read.cs

```diff
@@ -0,0 +1,52 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Microsoft.Win32.SafeHandles;
+using System;
+
+internal static partial class Interop
+{
+    internal static partial class Sys
+    {
+        /// <summary>
+        /// Reads a number of bytes from an open file descriptor into a specified buffer.
+        /// </summary>
+        /// <param name="fd">The open file descriptor to try to read from</param>
+        /// <param name="buffer">The buffer to read info into</param>
+        /// <param name="count">The size of the buffer</param>
+        /// <returns>
+        /// Returns the number of bytes read on success; otherwise, -1 is returned
+        /// Note - on fail. the position of the stream may change depending on the platform; consult man 2 read for more info
+        /// </returns>
+        internal static unsafe int Read (SafeFileHandle fd, byte* buffer, int count)
+        {
+            int bytes_read = -1;
+            IntPtr unmanaged_buffer = Marshal.AllocHGlobal (count);
+            bool has_ref = false;
+            try {
+                fd.DangerousAddRef (ref has_ref);
+                if (has_ref)
```

Comment:
`has_ref` will always be true if `fd.DangerousAddRef` succeeded so no need to check for it here. And the convention is to name it `release`.

Reply:
@luhenry ah, but `has_ref` will always be false if `fd.DangerousAddRef` failed :) I see that the common use of DangerousAddRef is to only use the success indicator on release, but that is not exclusive. In this case as we could be passing a recycled handle to a syscall, which cannot not be detected at any point later on as an invalid handle. So I am not convinced that I should remove this, and at worst it is harmless

Reply:
If `fd.DangerousAddRef` fails, it throws. The only use of `has_ref` is for the finally clause, since it will execute even in case of exceptions, and that's also why you pass `has_ref` by ref so it's set even in case of exception. What you have is just not how you use SafeHandle, so just look at  https://github.com/mono/mono/blob/master/mcs/class/System/System.Net.Sockets/Socket.cs#L187 to see an example on how you use it.

Reply:
@luhenry I missed that DangerousAddRef throws (the docs don't mention any exceptions for the method!!). You are totally right!